README.md
#  Flutter Floating Navigation Bar

:blue_heart:
<img width="450" height="900" src="https://github.com/tailoristic/floating_navigation_bar/blob/master/example/images/floating_navigation_bar.jpg?raw=true">

## ✨ EXAMPLE

```dart
      Scaffold(
      appBar: AppBar(
        title: Text(
          "Floating Navigation Bar"
        ),
      ),
      // INCLUDE LIST OF WIDGETS INSIDE BODY
      body: List<Widgets>, // [Home(), Search(), Cart(), Profile()]
      bottomNavigationBar: FloatingNavigationBar(
        backgroundColor: Colors.black87,
        iconColor: Colors.white,
        textStyle: TextStyle(
          color: Colors.white,
          fontSize: 14.0,
        ),
        iconSize: 20.0,
        items: [
          NavBarItems(icon: EvaIcons.homeOutline,title: "Home"),
          NavBarItems( icon: EvaIcons.search,title: "Search"),
          NavBarItems(
            icon: EvaIcons.shoppingCartOutline, title: "Cart"),
          NavBarItems(
            icon: EvaIcons.activity,title: "Profile"),
        ],
        onChanged: (value) {
            // USE YOUR STATE MANAGEMENT TECHNIQUE TO GET
            // AND CHANGE INDEX OF NAVIGATION BAR
        },
      ),
    );

```
Willing to contribute?
Go ahead and send pull request :grin:

Found bug or issues?:disappointed:
Open and issue on Github repo

## ⭐️ License

MIT License


CHANGELOG.md
## 0.0.1

* This is the initial release of Floating Navigation Bar.

## 0.0.2
* Partial Migration to null safety

## 0.0.3
* Added Indicator for active tab.

## 0.0.4
* Added more information in docs


## 0.0.5
* Added more information in docs

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:fnb_demo/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/floating_navigation_bar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:floating_navigation_bar/floating_navigation_bar.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/lib/drawer.dart
import 'package:eva_icons_flutter/eva_icons_flutter.dart';
import 'package:floating_navigation_bar/floating_navigation_bar.dart';
import 'package:flutter/material.dart';

class SideDrawer extends StatelessWidget {
  static int currentIndex = 0;

  static var pageList = <Widget>[
    Center(
      child: Text("Home Page"),
    ),
    Center(
      child: Text("Feed Page"),
    ),
    Center(
      child: Text("Cart Page"),
    ),
    Center(
      child: Text("Profile Page"),
    ),
  ];
  @override
  Widget build(BuildContext context) {
    return Drawer(
      elevation: 0.0,
      child: Container(
        color: Colors.black87,
        child: Column(
          children: [
            FloatingNavigationBar(
              backgroundColor: Colors.black87,
              iconColor: Colors.white,
              textStyle: TextStyle(
                color: Colors.white,
                fontSize: 14.0,
              ),
              iconSize: 20.0,
              items: [
                NavBarItems(
                  icon: EvaIcons.homeOutline,
                  title: "Home",
                ),
                NavBarItems(
                  icon: EvaIcons.search,
                  title: "Search",
                ),
                NavBarItems(
                  icon: EvaIcons.shoppingCartOutline,
                  title: "Cart",
                ),
                NavBarItems(
                  icon: EvaIcons.activity,
                  title: "Profile",
                ),
              ],
              onChanged: (_) {},
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/floating_navigation_bar.dart
import 'package:floating_navigation_bar/src/nav_bar_items.dart';
import 'package:flutter/material.dart';

class FloatingNavigationBar extends StatelessWidget {
  /// * TO CHANGE THE HEIGHT OF NAVIGATION BAR
  /// * PREFERRED HEIGHT IS BETWEEN [80-100]
  final double barHeight;

  /// * TO CHANGE THE WIDTH OF NAVIGATION BAR
  final double? barWidth;

  /// * FOR CHANGING NAV ITEMS ICON SIZE
  final double? iconSize;

  /// * For changing nav items icon color
  final Color? iconColor;

  /// * Background Color or floating navigation bar
  final Color? backgroundColor;

  //
  final TextStyle? textStyle;

  /// * List to tabs to display
  final List<NavBarItems> items;

  final ValueChanged<int> onChanged;

  /// * For Changing indicator color
  /// * ... height
  /// * ... width
  final Color indicatorColor;
  final double indicatorHeight;
  final double indicatorWidth;

  /// ! THIS WILL IDENTITFY WHICH TAB IS CURRENTLY ACTIVE
  /// ! USING THIS TO SHOW AND HIDE TAB INDICATOR
  static ValueNotifier<int> notifyIndex = ValueNotifier<int>(0);

  const FloatingNavigationBar({
    Key? key,
    required this.items,
    required this.onChanged,
    this.barHeight = 80.0,
    this.barWidth = 400.0,
    this.iconColor,
    this.iconSize,
    this.textStyle,
    this.backgroundColor,
    this.indicatorColor = Colors.black,
    this.indicatorHeight = 5.0,
    this.indicatorWidth = 8.0,
  })  : assert(
          (items.length < 5),
          "NavBarItems can't contain more than 4 itmes",
        ),
        assert(
          barHeight <= 100,
          "\n******\nHeight should be less than or equal to 100\n******\n",
        ),
        assert(
          indicatorWidth <= 15 || indicatorHeight <= 15,
          "\n******\n Too much height given to tab indicator \n******\n",
        ),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      height: barHeight,
      width: barWidth,
      margin: const EdgeInsets.symmetric(
        horizontal: 15.0,
        vertical: 20.0,
      ),
      padding: const EdgeInsets.all(5),
      decoration: BoxDecoration(
        color: backgroundColor,
        borderRadius: BorderRadius.circular(20.0),
      ),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceAround,
        children: List.generate(
          items.length,
          (i) => GestureDetector(
            onTap: () => onPressed(i),
            child: Container(
              width: 90.0,
              color: Colors.transparent,
              child: Column(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  Icon(
                    items[i].icon,
                    size: iconSize,
                    color: iconColor,
                  ),
                  Text(
                    items[i].title,
                    style: textStyle,
                    maxLines: 1,
                    overflow: TextOverflow.ellipsis,
                  ),
                  ValueListenableBuilder(
                    valueListenable: notifyIndex,
                    builder: (BuildContext context, int value, Widget? child) {
                      return Visibility(
                        visible: value == i,
                        child: Container(
                          height: indicatorHeight,
                          width: indicatorWidth,
                          padding: EdgeInsets.zero,
                          decoration: BoxDecoration(
                            borderRadius: BorderRadius.circular(5.0),
                            color: indicatorColor,
                          ),
                        ),
                      );
                    },
                  )
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }

  /// * FOR CHANGING NAVIGATION BAR  INDEX
  void onPressed(int index) {
    onChanged(index);
    notifyIndex.value = index;
  }
}


example/lib/main.dart
import 'package:eva_icons_flutter/eva_icons_flutter.dart';
import 'package:floating_navigation_bar/floating_navigation_bar.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Floating Navigation Bar',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: DemoScreen(),
    );
  }
}

class DemoScreen extends StatefulWidget {
  @override
  _DemoScreenState createState() => _DemoScreenState();
}

class _DemoScreenState extends State<DemoScreen> {
  int currentIndex = 0;

  var pageList = <Widget>[
    Center(
      child: Text("Home Page"),
    ),
    Center(
      child: Text("Feed Page"),
    ),
    Center(
      child: Text("Cart Page"),
    ),
    Center(
      child: Text("Profile Page"),
    ),
  ];
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.grey[140],
      appBar: AppBar(
        elevation: 0.0,
        toolbarHeight: 70.0,
        backgroundColor: Colors.grey[200],
        title: Text(
          "Floating Navigation Bar",
          style: TextStyle(
            color: Colors.black,
          ),
        ),
      ),
      body: pageList[currentIndex],
      bottomNavigationBar: FloatingNavigationBar(
        backgroundColor: Colors.black87,
        barHeight: 80.0,
        barWidth: MediaQuery.of(context).size.width - 40.0,
        iconColor: Colors.white,
        textStyle: TextStyle(
          color: Colors.white,
          fontSize: 14.0,
        ),
        iconSize: 20.0,
        indicatorColor: Colors.white,
        indicatorHeight: 5,
        indicatorWidth: 14.0,
        items: [
          NavBarItems(
            icon: EvaIcons.homeOutline,
            title: "Home",
          ),
          NavBarItems(
            icon: EvaIcons.search,
            title: "Search",
          ),
          NavBarItems(
            icon: EvaIcons.shoppingCartOutline,
            title: "Cart",
          ),
          NavBarItems(
            icon: EvaIcons.activity,
            title: "Profile",
          ),
        ],
        onChanged: (value) {
          currentIndex = value;
          setState(() {});
        },
      ),
    );
  }
}
